Feature/zcs 11179

SuryawanshiAmol opened the pull request.

SuryawanshiAmol committed:
ZCOMT-2586 - Compile the nginx code in branch feature/ZCS-11179

- There appears to be a significant change in request member "cookie" that must be handled. Code is commented in order to move compilation forward and detect any errors that arise. ( Previously it was ngx_array_t and now its changed to ngx_table_elt_t which don't have  members elts & nelts).

SuryawanshiAmol committed:
ZCOMT-2586 - Compile the nginx code in branch feature/ZCS-11179

- Updated nginx version

SuryawanshiAmol committed:
ZCOMT-2586 - Compile the nginx code in branch feature/ZCS-11179

- There appears to be a significant change in request member "cookie" that must be handled. Code is commented in order to move compilation forward and detect any errors that arise. ( Previously it was ngx_array_t and now its changed to ngx_table_elt_t which don't have  members elts & nelts).
- Fixed the problem by breaking the external for loop as its not needed to loop over as cookie data may come in string

SuryawanshiAmol committed:
ZCOMT-2586 - Compile the nginx code in branch feature/ZCS-11179

- Modified some logging statements in-order to differentiate it from original logs. This changes will be reverted once testing done.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Zimbra/packages?pullRequest=194) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Zimbra/packages?pullRequest=194) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: umagmrit<br/>:x: SuryawanshiAmol<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Zimbra/packages?pullRequest=194) it.</sub>

SuryawanshiAmol committed:
ZCOMT-2586 - Compile the nginx code in branch feature/ZCS-11179

- Modified some logging statements in-order to differentiate it from original logs. This changes will be reverted once testing done.

SuryawanshiAmol committed:
ZCOMT-2586 - Compile the nginx code in branch feature/ZCS-11179

- Minor change. Initialized return bool variable ngx_flag_t to 0.

SuryawanshiAmol committed:
ZCS-13998 : NGINX - Customization code separation from repo

- Created the patch files to add zimbra specific customization on the top of the public nginx repository stable-1.24.0

SuryawanshiAmol committed:
ZCS-13998 : NGINX - Customization code separation from repo

- Changed in order to work the customization for the any nginx version got from public repository.

ZCS-13998 : NGINX - Customization code separation from repo

- Added patch files in order to apply zimbra specific customizations

ZCS-13998 : NGINX - Customization code separation from repo

- Changes to create build with patches

Update Makefile

ZCS-13998 : NGINX - Customization code separation from repo

- Update build file

ZCS-13998 : NGINX - Customization code separation from repo

- customization related changes

ZCS-13998 : NGINX - Customization code separation from repo

- build failure fix

ZCS-13998 : NGINX - Customization code separation from repo

ZCS-13998 : NGINX - Customization code separation from repo

- build failure fixes

ZCS-13998 : NGINX - Customization code separation from repo

- fix

ZCS-13998 : NGINX - Customization code separation from repo

- fix

ZCS-13998 : NGINX - Customization code separation from repo

- compilation fixes

ZCS-13998 : NGINX - Customization code separation from repo

- fixed compilation

ghost commented:
When I try to compile NGinx for Ubuntu 20, it works, but there is an error when creating the source package. This means the binary package works, but the source package is not created. I am not sure if we need to provide the source package (to comply with the license) since we already release our patch files. So perhaps this step can be removed.

```
make[1]: Leaving directory '/tmp/packages/thirdparty/nginx/build/UBUNTU20_64/zimbra-nginx'
 dpkg-source -irpm -b .
dpkg-source: info: using source format '3.0 (quilt)'
dpkg-source: info: building zimbra-nginx using existing ./zimbra-nginx_1.24.0.orig.tar.gz
dpkg-source: info: using patch list from debian/patches/series
dpkg-source: error: cannot represent change to objs/addon/nviennot-nginx-tcp-keepalive/ngx_http_tcp_keepalive_module.o: binary file contents changed
dpkg-source: error: add objs/addon/nviennot-nginx-tcp-keepalive/ngx_http_tcp_keepalive_module.o in debian/source/include-binaries if you want to store the modified binary in the debian tarball
dpkg-source: error: cannot represent change to objs/addon/sso/ngx_http_zm_sso_module.o: binary file contents changed
dpkg-source: error: add objs/addon/sso/ngx_http_zm_sso_module.o in debian/source/include-binaries if you want to store the modified binary in the debian tarball
dpkg-source: error: cannot represent change to objs/addon/throttle/ngx_mail_throttle_module.o: binary file contents changed
dpkg-source: error: add objs/addon/throttle/ngx_mail_throttle_module.o in debian/source/include-binaries if you want to store the modified binary in the debian tarball
dpkg-source: error: cannot represent change to objs/addon/upstreamzmauth/ngx_http_upstream_zmauth_module.o: binary file contents changed
dpkg-source: error: add objs/addon/upstreamzmauth/ngx_http_upstream_zmauth_module.o in debian/source/include-binaries if you want to store the modified binary in the debian tarball
dpkg-source: error: cannot represent change to objs/addon/zmauth/ngx_mail_zmauth_module.o: binary file contents changed
dpkg-source: error: add objs/addon/zmauth/ngx_mail_zmauth_module.o in debian/source/include-binaries if you want to store the modified binary in the debian tarball
dpkg-source: error: cannot represent change to objs/nginx: binary file contents changed
dpkg-source: error: add objs/nginx in debian/source/include-binaries if you want to store the modified binary in the debian tarball
dpkg-source: warning: executable mode 0755 of 'objs/nginx' will not be represented in diff
...
repeats 100 or so more times
```

barrydegraaff commented:
Please explain how to build using this change.

silentsakky commented on thirdparty/nginx/patches/nginx_README.patch:
```diff
@@ -0,0 +1,14 @@
++
++This is the official repository for building out the third party dependency nginx for Zimbra Collaboration Suite 8.7 and later.
++
++Issues should be reported via [Zimbra's bugzilla](https://bugzilla.zimbra.com)
```
bugzilla is no more :)

SuryawanshiAmol replied:
Yes. Will add new Jira link here.

barrydegraaff reviewed (approved):
It is a little difficult to verify the patches from https://github.com/Zimbra/packages/pull/194 are the same as what is in https://github.com/Zimbra/nginx/pull/5 because we do not apply our changes on the upstream nginx source. Instead we copy upstream and apply them along with all our other nginx modules and then also remove and move files.

After manually moving and removing some files,  I can confirm that this https://github.com/Zimbra/packages/pull/194 produces the same as https://github.com/Zimbra/nginx/pull/5

The source package throws many errors for me such as 
```
dpkg-source: error: cannot represent change to objs/addon/nviennot-nginx-tcp-keepalive/ngx_http_tcp_keepalive_module.o: binary file contents changed
```

But since that source package is AFAIK not used, the PR is _good enough_.

Verified using:

```
#!/bin/bash

cd /tmp
rm packages zimbra-* -Rf

apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 9BE6ED79

rm /etc/apt/sources.list.d/zimbra.list
cat >> /etc/apt/sources.list.d/zimbra.list << EOF
deb [arch=amd64] https://repo.zimbra.com/apt/87 focal zimbra
deb-src [arch=amd64] https://repo.zimbra.com/apt/87 focal zimbra
deb [arch=amd64] https://repo.zimbra.com/apt/1000 focal zimbra
deb-src [arch=amd64] https://repo.zimbra.com/apt/1000 focal zimbra
deb [arch=amd64] https://repo.zimbra.com/apt/1000-ne focal zimbra
deb-src [arch=amd64] https://repo.zimbra.com/apt/1000-ne focal zimbra
EOF

apt-get clean
apt update

apt-get source gkrellweather
apt-get build-dep gkrellweather

apt -y install m4 libpcre3-dev build-essential git binutils lintian debhelper dh-make devscripts libz-dev quilt

GIT_SSH_COMMAND="ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no" git clone git@github.com:Zimbra/packages.git
GIT_SSH_COMMAND="ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no" git clone git@github.com:Zimbra/zimbra-build.git
GIT_SSH_COMMAND="ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no" git clone git@github.com:Zimbra/zimbra-package-stub.git

cd zimbra-package-stub
git checkout develop #ubuntu 20 support
cd ..

cd packages
git checkout feature/ZCS-11179
git submodule update --init --recursive --remote

cd /tmp/packages/thirdparty/nginx/
cd nginx
#git checkout feature/ZCS-11179
rm -Rf nginx
cd ..
make

sleep 10

make build

echo "Compiled version:"
/tmp/packages/thirdparty/nginx/build/UBUNTU20_64/zimbra-nginx/objs/nginx -V

```

SuryawanshiAmol committed:
ZCS-13998 : NGINX - Customization code separation from repo

- Some review comments

SuryawanshiAmol committed:
ZCS-13998 : NGINX - Customization code separation from repo

- Some review comments

SuryawanshiAmol committed:
ZCS-14068 : NGINX - Update review comments and documentation

- Some review comments

SuryawanshiAmol committed:
ZCS-14068 : NGINX - Update review comments and documentation

- Some review comments

umagmrit committed:
ZCS-11179:Updated changelog for nginx

umagmrit committed:
ZCS-11179:Updated proxy-components
